Feature: support endpoint name grouping by OpenAPI definitions. (#7130)
apache/skywalking: docs/en/setup/backend/endpoint-grouping-rules.md

          type: string
          description: Product description
      required:
        - id
        - name
    RelatedProducts:
      type: object
      description: Related Products
      properties:
        id:
          type: integer
          format: int32
          description: Product id
        relatedProducts:
          type: array
          description: List of related products
          items:
            $ref: "#/components/schemas/Product"

```

Here are some cases:

   | Incoming Endpiont | Incoming Service | x-sw-endpoint-name-match-rule | x-sw-endpoint-name-format | Matched | Grouping Result |
   |-----|-----|-----|-----|-----|-----|
   | GET:/products | serviceB | default | default | true | GET:/products |
   | GET:/products/123 | serviceB | default |default |  true | GET:/products{id} |
   | GET:/products/asia/cn | serviceB | default | default | true | GET:/products/{region}/{country} |
   | GET:/products/123/abc/efg | serviceB | default |default |  false | GET:/products/123/abc/efg | 
   | \<GET\>:/products/123 | serviceB | default | default | false | \<GET\>:/products/123|
   | GET:/products/123 | serviceC | default | default | false | GET:/products/123 |

>    | \<GET\>:/products/123 | serviceB | \<\${METHOD}\>:\${PATH} | \<\${METHOD}>:\${PATH} | true | \<GET\>:/products/{id} |

Reviewer: See 
```suggestion
   | \<GET\>:/products/123 | serviceB | \<\${METHOD}\>:\${PATH} | \<\${METHOD}\>:\${PATH} | true | \<GET\>:/products/{id} |
```

>    | GET:/products/123 | serviceB | default | ${PATH}:\<\${METHOD}\> | true | /products/{id}:\<GET\> |

Reviewer: See
```suggestion
   | GET:/products/123 | serviceB | default | \${PATH}:\<\${METHOD}\> | true | /products/{id}:\<GET\> |
```

have an endpoint name like `prod/{order-id}`. This is the feature of parameterized endpoint grouping designed for.

If the incoming endpoint name hit the rules, SkyWalking will grouping the endpoint by rules.

SkyWalking provides 2 ways to support endpoint grouping:
1. Endpoint name grouping by OpenAPI definitions.
2. Endpoint name grouping by custom configuration.

The 2 grouping features can work together in sequence.
## Endpoint name grouping by OpenAPI definitions
The OpenAPI definitions are the documents based on The [OpenAPI Specification (OAS)](https://github.com/OAI/OpenAPI-Specification) which used to define a standard, language-agnostic interface for HTTP APIs.
SkyWalking now support `OAS v2.0+)`, could parse the documents `(yaml)` and build the grouping rules from them automatically.

### How to use
1. Add some `Specification Extensions` for SkyWalking in the OpenAPI definition documents:<br />
   `${METHOD}` is a reserved placeholder which represents the HTTP method eg. `POST/GET...` <br />
   `${PATH}` is a reserved placeholder which represents the path eg. `/products/{id}`.

   | Extension Name | Required | Description | Default Value |
   |-----|-----|-----|-----|

>    | x-sw-service-name | true | The service name to which these endpoints belong | |

Reviewer 2: We should make this optional, and the default should be the folder name. The reason I am asking this is, once all things are not required, `x-sw-endpoint-name-match-rule`'s default value matched our ALS rule(@kezhenxu94 it is matched, right?), we could only ask for standard OpenAPI spec file in the right folder, nothing more.

Reviewer: Yes they matched

Author: make sense

Reviewer 2: > But if the folder is nested,

What do you mean `folder nested`? I think you should require a folder as the service name, but could use `x-sw-service-name` to override.

Reviewer: The folder can be any number depth

```
a
 |--b
 |--|--c
 |--|--|--d
```

Author: A recommend practice from OAS is `A good rule of thumb is to use the natural hierarchy present in URLs to build your file structure. For example, put all routes starting with /users (like /users and /users/{id}) in the same file (think of it as a “sub-API”).`

Author: I think using the folder name that the file belong is enough.

Reviewer 2: Read https://github.com/apache/skywalking/pull/7130#discussion_r654422789

Author: done

# Group Parameterized Endpoints
In most cases, the endpoint should be detected automatically through the language agents, service mesh observability solution,
or configuration of meter system.

There are some special cases, especially when people use REST style URI, the application codes put the parameter in the endpoint name,
such as putting order id in the URI, like `/prod/ORDER123` and `/prod/ORDER123`. But logically, people expect they could
have an endpoint name like `prod/{order-id}`. This is the feature of parameterized endpoint grouping designed for.

If the incoming endpoint name hit the rules, SkyWalking will grouping the endpoint by rules.

SkyWalking provides 2 ways to support endpoint grouping:
1. Endpoint name grouping by OpenAPI definitions.
2. Endpoint name grouping by custom configuration.

The 2 grouping features can work together in sequence.
## Endpoint name grouping by OpenAPI definitions

> The OpenAPI definitions are the documents based on The [OpenAPI Specification (OAS)](https://github.com/OAI/OpenAPI-Specification) which used to define a standard, language-agnostic interface for HTTP APIs.

Reviewer 2: ```suggestion
The OpenAPI definitions are the documents based on The [OpenAPI Specification (OAS)](https://www.openapis.org/) which used to define a standard, language-agnostic interface for HTTP APIs.
```

Once we say a spec, we should link to their website rather that GitHub repo. GitHub is for developers.

SkyWalking now support `OAS v2.0+`, could parse the documents `(yaml)` and build the grouping rules from them automatically.

### How to use
1. Add some `Specification Extensions` for SkyWalking config in the OpenAPI definition documents, otherwise, all configs are default:<br />
   `${METHOD}` is a reserved placeholder which represents the HTTP method eg. `POST/GET...` <br />
   `${PATH}` is a reserved placeholder which represents the path eg. `/products/{id}`.

   | Extension Name | Required | Description | Default Value |
   |-----|-----|-----|-----|

>    | x-sw-service-name | false | The service name to which these endpoints belong | The directory name which the definition documents belong to|

Reviewer 2: What happens in no `x-sw-service-name` and `yaml` file is located in the `openapi-definitions` directory directly.

Author: `openapi-definitions` would be the service name, should throw an exception?

   | x-sw-endpoint-name-match-rule | false | The rule used to match the endpoint.| `${METHOD}:${PATH}` |
   | x-sw-endpoint-name-format | false | The endpoint name after grouping.| `${METHOD}:${PATH}` |

   These extensions are under `OpenAPI Object`. For example, the document below has a full custom config:

``` yaml
openapi: 3.0.0
x-sw-service-name: serviceB
x-sw-endpoint-name-match-rule: "<${METHOD}>:${PATH}"

> x-sw-endpoint-name-format: "<${METHOD}>:${PATH}"

Reviewer 2: Let's keep these as same as default value, to avoid any confusion. 

@wankai123 We also need an issue and another PR to fix current operation name in plugins. You should fix those after this gets merged.

info:
  description: OpenAPI definition for SkyWalking test.
  version: v2
  title: Product API
  ...
```

   We highly recommend using the default config, the custom config (`x-sw-endpoint-name-match-rule/x-sw-endpoint-name-format`) would be considered as part of the match rules (regex pattern).
   We provide some cases in `org.apache.skywalking.oap.server.core.config.group.openapi.EndpointGroupingRuleReader4OpenapiTest`, you could validate your custom config as well.

> 1. Put the OpenAPI definition documents into directory `openapi-definitions`, SkyWalking could read all documents or documents in this subDirectorys from it, you can organize these documents by yourself. Recommend using the service name as the subDirectory name then you are not necessary to set `x-sw-service-name`. For example:

Reviewer 2: ```suggestion
1. All OpenAPI definition documents should be located in the `openapi-definitions` directory, with at most two levels directories. Recommend using the service name as the subDirectory name then you are not necessary to set `x-sw-service-name`. For example:
```

  ```
├── openapi-definitions
│   ├── serviceA
│   │   ├── customerAPI-v1.yaml
│   │   └── productAPI-v1.yaml
│   └── serviceB
│       └── productAPI-v2.yaml
```

> 3. Turn the feature on by setting the `Core Module` configuration `${SW_CORE_ENABLE_ENDPOINT_NAME_GROUPING_BY_OPAENAPI:true}`

Reviewer 2: As no side effect, keeping this open in default should be better?

        "400":
          description: Invalid product id

components:
  schemas:
    Product:
      type: object
      description: Product id and name
      properties:
        id:
          type: integer
          format: int64
          description: Product id
        name:
          type: string
          description: Product name
      required:
        - id
        - name
    ProductDetails:
      type: object
      description: Product details
      properties:
        id:
          type: integer
          format: int64
          description: Product id
        name:
          type: string
          description: Product name
        description:
          type: string
          description: Product description
      required:
        - id
        - name
    RelatedProducts:
      type: object
      description: Related Products
      properties:
        id:
          type: integer
          format: int32
          description: Product id
        relatedProducts:
          type: array
          description: List of related products
          items:
            $ref: "#/components/schemas/Product"

```

Here give some scenario we might use:

> 1. Only set the `x-sw-service-name`, `x-sw-endpoint-name-match-rule` and `x-sw-endpoint-name-format` are default:

Reviewer 2: I think these 3 scenarios are strange to describe.

``` yaml
openapi: 3.0.0
x-sw-service-name: serviceB

info:
  description: OpenAPI definition for SkyWalking test.
  version: v2
  title: Product API
  ...
```
2. Set the `x-sw-service-name` , `x-sw-endpoint-name-match-rule` and `x-sw-endpoint-name-format` :
``` yaml
openapi: 3.0.0
x-sw-service-name: serviceB
x-sw-endpoint-name-match-rule: "<${METHOD}>:${PATH}"
x-sw-endpoint-name-format: "<${METHOD}>:${PATH}"

info:
  description: OpenAPI definition for SkyWalking test.
  version: v2
  title: Product API
  ...
```
3. Set the `x-sw-service-name` , `x-sw-endpoint-name-match-rule` and `x-sw-endpoint-name-format` :
``` yaml
openapi: 3.0.0
x-sw-service-name: serviceB
x-sw-endpoint-name-match-rule: "<${METHOD}>:${PATH}"

> x-sw-endpoint-name-format: "<${METHOD}>:${PATH}"

Reviewer 2: Move this as an example under the table should be good enough.

![image](https://user-images.githubusercontent.com/5441976/122521441-658e1d80-d047-11eb-8db3-4694f56981d3.png)

info:
  description: OpenAPI definition for SkyWalking test.
  version: v2
  title: Product API
  ...
```

   | Incoming Endpiont | Incoming Service | x-sw-endpoint-name-match-rule | x-sw-endpoint-name-format | Matched | Grouping Result |
   |-----|-----|-----|-----|-----|-----|
   | GET:/products | serviceB | default | default | true | GET:/products |
   | GET:/products/123 | serviceB | default |default |  true | GET:/products{id} |
   | GET:/products/asia/cn | serviceB | default | default | true | GET:/products/{region}/{country} |
   | GET:/products/123/abc/efg | serviceB | default |default |  false | GET:/products/123/abc/efg | 
   | \<GET\>:/products/123 | serviceB | default | default | false | \<GET\>:/products/123|
   | GET:/products/123 | serviceC | default | default | false | GET:/products/123 |
   | \<GET\>:/products/123 | serviceB | \<\${METHOD}\>:\${PATH} | \<\${METHOD}>:\${PATH} | true | \<GET\>:/products/{id} |
   | GET:/products/123 | serviceB | default | ${PATH}:\<\${METHOD}\> | true | /products/{id}:\<GET\> |

>    | /products/123:\<GET\> | serviceB | ${PATH}:\<\${METHOD}\> | default | true | GET:/products/{id} |

Reviewer 2: This table is better than the above 3 cases.

>    | /products/123:\<GET\> | serviceB | ${PATH}:\<\${METHOD}\> | default | true | GET:/products/{id} |

Reviewer: I'd rather put these GET:/products, etc. etc. into `` so you don't need to escape the characters, which is error prone

Reviewer 2: I am wondering the point of `\` too

Author: to escape the markdown keyword...  but if I put them into `` ,  the style (color) would differ from others.

Reviewer 2: Could you put a screenshot here to demonstrate what will be seen?

Author: ![image](https://user-images.githubusercontent.com/16773043/122528994-8490ad80-d04f-11eb-8c07-d7d4ddccea2e.png)
would like this

Reviewer 2: @kezhenxu94 From the preview, it seems better to keep the current way?

Reviewer: - `$`, `<`, and `>` has no special meanings in GitHub markdown, they may have in other extended markdown engine.

- You don't have consistent style in terms of quoting in backticks or not, for example you quote them here 
<img width="1065" alt="Screen Shot 2021-06-18 at 17 54 24" src="https://user-images.githubusercontent.com/15965696/122543360-46e75100-d05e-11eb-8d81-bbe8a1806e58.png">
but not in this table

- If you choose not to quote in `` but with `\` to escape those characters, see 

- There are cases when users may look into the source codes and copy the example with those `\`, which doesn't work.

Author: make sense, I have put all path/pattern string into ``.

> <br />

Reviewer 2: ```suggestion

```
